### PR TITLE
Add Java DSL guide

### PR DESCRIPTION
**Pull request by chrisjs**
Resolves https://github.com/spring-cloud/spring-cloud-dataflow/issues/3212

### CODE REVIEW

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md [hunk 1]
```diff
@@ -4,4 +4,255 @@ title: 'Java DSL'
...
+Instead of using the shell to create and deploy streams, you can use the Java-based DSL provided by the `spring-cloud-dataflow-rest-client` module.
+The Java DSL is a convenient wrapper around the `DataFlowTemplate` class that enables creating and deploying streams programmatically.
+
+To get started, you need to add the following dependency to your project, as follows:
+
+```xml
+<dependency>
+	<groupId>org.springframework.cloud</groupId>
+	<artifactId>spring-cloud-dataflow-rest-client</artifactId>
+	<version>{project-version}</version>
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: Does `{project-version}` gets replaced when we do the Microsite build?

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md [hunk 2]
```diff
@@ -4,4 +4,255 @@ title: 'Java DSL'
...
+</dependency>
+```
+
+<!--NOTE-->
+
+A complete sample can be found in the [Spring Cloud Data Flow Samples Repository](https://github.com/spring-cloud/spring-cloud-dataflow-samples).
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: It'd be nice to link to that sample directly as opposed to the root folder. See: https://docs.spring.io/spring-cloud-dataflow-samples/docs/current/reference/htmlsingle/#_java_dsl

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md [hunk 3]
```diff
@@ -4,4 +4,255 @@ title: 'Java DSL'
...
+The properties in [DataFlowClientProperties](https://github.com/spring-cloud/spring-cloud-dataflow/blob/master/spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/config/DataFlowClientProperties.java) can be used to configure the connection to the Data Flow server.
+The common property to start using is `spring.cloud.dataflow.client.uri`
+
+Consider the following example, using the `definition` style.
+
+```java
+URI dataFlowUri = URI.create("http://localhost:9393");
+DataFlowOperations dataFlowOperations = new DataFlowTemplate(dataFlowUri);
+dataFlowOperations.appRegistryOperations().importFromResource(
+                     "https://bit.ly/Darwin-SR3-stream-applications-rabbit-maven", true);
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: Let's switch this to Einstein latest.

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md [hunk 4]
```diff
@@ -4,4 +4,255 @@ title: 'Java DSL'
...
+The Java DSL offers two styles to create Streams.
+
+- The `definition` style keeps the feel of using the pipes and filters textual DSL in the shell.
+  This style is selected by using the `definition` method after setting the stream name - for example, `Stream.builder(dataFlowOperations).name("ticktock").definition(<definition goes here>)`.
+- The `fluent` style lets you chain together sources, processors, and sinks by passing in an instance of a `StreamApplication`.
+  This style is selected by using the `source` method after setting the stream name - for example, `Stream.builder(dataFlowOperations).name("ticktock").source(<stream application instance goes here>)`.
+  You then chain together `processor()` and `sink()` methods to create a stream definition.
+
+To demonstrate both styles, we include a simple stream that uses both approaches.
+A complete sample for you to get started can be found in the [Spring Cloud Data Flow Samples Repository](https://github.com/spring-cloud/spring-cloud-dataflow-samples).
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: Here, again, we could link to the exact sample.

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md [hunk 5]
```diff
@@ -4,4 +4,255 @@ title: 'Java DSL'
...
+
+In order to create and deploy your streams, you need to make sure that the corresponding apps have been registered in the DataFlow server first.
+Attempting to create or deploy a stream that contains an unknown app throws an exception.
+You can register your application by using the `DataFlowTemplate`, as follows:
+
+<!--END_NOTE-->
+
+```java
+dataFlowOperations.appRegistryOperations().importFromResource(
+            "https://bit.ly/Darwin-SR3-stream-applications-rabbit-maven", true);
```
Comment (sabbyanandan) [inline content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: Same, Einstein apps, please.

File: content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md (hunk 1 above)
Comment (markpollack) [reply content/documentation/pages/5-feature-guides/1-streams/7-java-dsl.md]: Need to be `%project-version%.  Working on getting that behaving as it did before with the maven build.  Can be part of another cleanup pass for variables.

**Comment by markpollack** [comment]
merged as ddd68f7cbc5d9d371b6be6b9b1c45444b5f9d1c1

**Pull request closed without merging**